### ISSUE TITLE
Add support for Marlin firmware

### ISSUE THREAD

**Issue by breiler** [winder/Universal-G-Code-Sender#1080]
## Feature request
Add support for Marlin controllers: http://marlinfw.org/

It may not be the first alternative for CNC machines as it's mostly focused on 3D printing. But the firmware supports commands for controlling spindles or lasers. The Marlin community is large and their controllers are widely available and also runs on a plain Arduino Mega. Could be a great alternative for DIY hackers.

These features needs to be implemented:
[ ] Streaming file, pausing and stopping
[ ] Processing of Gcode states
[ ] Status report for machine coordinates
[ ] Status report for work coordinates
[ ] Jogging
[ ] Homing
[ ] Firmware settings

**Comment by vMeph** [comment]
that would be lovely to have support for marlin
@breiler one feature i would love to see in UGS is the ability to import .brb files directly from autodesk eagle into UGS  like ChiliPeppr uses
https://github.com/chilipeppr/widget-eagle
would that be possible at some point?

**Comment by breiler** [comment]
It's going slow... :(

I might get a couple of hours next week to work on it. But it requires a lot of work to get all features in place.

**Comment by danriches** [comment]
I totally understand, if I wasn't so ridiculously busy at work I'd learn java and have a go at helping out, but unfortunately I've a very short deadline to produce a piece of hardware and being the sole hardware developer it's overwhelming :(

Great work by the way, love what you've done so far and it's great with my mini laser engraver!

Dan

**Comment by andy-hnq** [comment]
I've done a very bare bones version of the stuff needed to make UGS work with Marlin.

https://github.com/andy-hnq/Universal-G-Code-Sender/tree/marlin

As far as I can tell the main challenges (apart from me being a newbie both at Marlin and UGS) seem to be...
- Marlin doesn't want to reply to status queries etc. while it is in the middle of a stepper move. 
- It lacks any single character commands so the way that UGS bypasses the command queue for status updates, pauses etc doesn't work with Marlin.
- I don't have enough knowledge of Marlin's command buffering to get it working when not in single step mode

What works?
- Classic UI
- Jogging
- manual entry of G code
- Position reporting (at least when Marlin is able to answer)
- My CNC milling test job in single step mode (and I have cut metal with it)
- M0 in the G code to pause execution and the UGS pause/resume button to resume it

What doesn't work or is flaky?
- Some commands sent by UGS seem to be skipped when used without single step mode so don't use this for cutting metal
- origin/axis reset (close and open the connection will do it as a workaround)
- position update polling sometimes doesn't work on reopening a connection (you should check whether the update messages are appearing before starting a job)
- pause/resume button sometimes has the wrong label (but seems to do the right thing)
- position updates produce stuff in the window regardless of verbosity setting
- status shown on the GUI can go idle intermittently when a job is still running
- I haven't tested manually pausing from the GUI, but I expect problems
- probably almost everything else that is firmware dependent :)

What am I using?
- Windows 10 laptop
- JDK 1.8.0_121
- I'm using Marlin bugfix-2.0.x now but started with 1.1.x
- EMERGENCY_PARSER enabled (required for M0 pause/resume)
- Arduino Mega 2560 connected via USB
- 115200 baud
- Classic UI
- single step mode
- status polling enabled with 5000ms interval (the default is 200ms)

Use at your own risk. No warranty!

**Comment by breiler** [comment]
Awesome, I'll try to free up some time and have a look. 

I also made an attempt to get this working. But as you mentioned, real time commands would have been really helpful. Also, I couldn't get the machine coordinates or the work coordinate offsets to work.

**Comment by andy-hnq** [comment]
Cheers. I have not attempted to deal with machine vs work coords. At the moment wherever the cutter is when you connect is 0,0 so I just start from wherever I am or else jog to where I want to be and then disconnect and reconnect to zero the coordinates. Crude I know but it works for me.

**Comment by andy-hnq** [comment]
Having read a bit more about Marlin's command buffer I tried a change to limit the outstanding commands to 10 in an attempt to make it less flaky when not in single step mode.

I thought that had fixed it. My test job didn't make any obvious shortcut moves... or so I thought.

After running a few times and half noticing that one move looked not quite right I finally managed to catch it in the act of being wrong. It doesn't have the same issue in single step mode.

So it seems improved but it's still not right. It's now missing fewer steps but it's not fixed.

I also messed up the git branch so had to recreate it.

**Comment by andy-hnq** [comment]
I just pushed a commit.

I've tried some more debugging of the non-single step command loss and added some logging to check what is happening.

I confirmed that my 10 outstanding command limit seems to work as intended. I've changed it so that it is no longer bending the communicator class pause semantics but that wasn't the cause of the problem.

I discovered that the M108 command was incorrectly being sent twice and fixed that.

I also noticed that the first skipped command was sent immediately before the M108 so I have altered the communicator so that it immediately pauses sending when an M0 is sent.

My test job now gets further before hitting an issue but then it not only skips a move but seems to mess up the coordinate position too - almost as if the move happened... except it didn't so the finishing point is offset by the distance of the lost move but M114 reports 0,0 at the end of the job.

My test job still works OK in single step mode.

**Comment by carneeki** [comment]
Glad to progress on this - I'd love to ditch OctoPrint for my Trinus (which I have put Marlin on after wiping the factory firmware).

Will be happy to help test for printing.

**Comment by andy-hnq** [comment]
Cheers. If you are happy to build from the sources using 'mvn package' then you should be good to go.

My machine is a CNC miller with steppers on X and Y and nothing else, so a printer is quite a different beast.

At the moment I'm thinking that the non-single step flakiness is primarily related to M0/M108 pause/resume. I'm intending to try a change to make UGS wait for the active command list to drain before sending an M0 that it finds in the g code script but haven't had time to try that yet.

**Comment by andy-hnq** [comment]
I had another try at fixing the lost moves issue. No joy I'm afraid.

My UGS branch now delays sending M0 until the active command list is empty. 

My test job has an M0 near the start and one near the end - these are prompts for me to lower the cutter at the start and raise it at the end. I had thought that the early sending of the second M0 was causing the problem but it seems not - I took the second one out of the script and it made no difference. It seems to be the first M0 that is causing the problem (the job works fine if there are no M0 commands).

In desperation I tried disabling EMERGENCY_PARSER and reflashing the Arduino. This causes it to complain that M0 is not recognised. UGS still sends M0 and M108 and the job runs OK (effectively my changes to the M0 handling mean that UGS is implementing the pause instead of Marlin). 

I'm out of ideas just now for ways to get M0 pause/resume to work reliably when not in single step mode and with EMERGENCY_PARSER enabled in the Marlin build.

**Comment by mron** [comment]
What are the minimum commands/behavior that would need to be implemented in Marlin to make this UGS work?

**Comment by andy-hnq** [comment]
> 
> 
> What are the minimum commands/behavior that would need to be implemented in Marlin to make this UGS work?

I guess it depends on your machine and the work you want to do. I can't answer that for everyone but I am still using my branch and it's doing the job for me just fine but my setup is very simple - X and Y steppers only, 'classic UI' only. I have not done any more on this since my previous post but if anyone is waiting on something that is easily do-able then please say so and I will give it a try. I have no way of testing stuff that my mill can't do though (e.g. limit switches)

**Comment by andy-hnq** [comment]
Sorry just re-read your question and realised you were asking about Marlin changes not UGS changes. In general the issue seems to be that Marlin is not so strong on async behaviour, for example being able to find out the current coordinates in the middle of a move, pause in the middle of a move or be able to reliably queue a bunch of commands and work through the queue with the same end result as if the commands were sent one at a time. I am definitely no Marlin expert though so it's quite possible that I'm just doing it wrong. If so then constructive input would be very welcome :)

**Comment by mron** [comment]
Thanks for replying. I'm not a Marlin expert, but I have modified it a few times for special projects. If you can give me a spec for new behavior and/or commands I'll see if I can implement them. List them in order of importance. I have added emergency codes in the past.

**Comment by mron** [comment]
I'm asking for "better" support, not perfect. I'm not an expert in Marlin or CNC.

**Comment by andy-hnq** [comment]
So in terms of commands, this is probably a good place to start (com.willwinder.universalgcodesender.GrblUtils) ...

    /**
     * Grbl commands
     */
    // Real time
    public static final byte GRBL_PAUSE_COMMAND = '!';
    public static final byte GRBL_RESUME_COMMAND = '~';
    public static final byte GRBL_STATUS_COMMAND = '?';
    public static final byte GRBL_DOOR_COMMAND = (byte)0x84;
    public static final byte GRBL_JOG_CANCEL_COMMAND = (byte)0x85;
    public static final byte GRBL_RESET_COMMAND = 0x18;

These are single byte 'commands' with GRBL and work immediately and regardless of whether the controller is busy or idle. The commands don't necessarily need to be identical in Marlin... provided that there is a way to do the same thing.

Priority-wise, I'd say status, pause/resume and then reset. I have never used the other two AFAIK.

As I've said in the thread above, the other thing I could not get working properly was the ability to queue and execute multiple commands reliably. For my own machine sending one command at a time doesn't make it do the job less well or even slow things down as far as I can tell. I assume that the capability to queue multiple commands is there for a reason... so knowing the 'use case' that requires this would be good.

**Comment by mron** [comment]
I’ll look at the status command and see what’s returned. Pause and resume will be trickier. The machine will need to decelerate and accelerate properly to keep the step count. Queuing commands might have something to do with motion planning. I’ve never looked into that part of the code.

**Comment by mron** [comment]
Hi Andy.
I’m working on getting this to work without losing steps. A few days before I started an identical projected was completed. This one uses quickstep(), so it will lose steps, but it’s functional.

I don’t know if any of it will be folded into Marlin, but it’s a good start.

Ron

Here’s the link:

https://github.com/fedetony/Marlin/tree/bugfix2.0.xautoreport

**Comment by mron** [comment]
I'm still working on the M0 feed hold. I have some questions. When M0 ( ! ) is issued and later resumed, where is the previous state kept?

Since Marlin queues up commands, does UGS know what commands have been completed? How do the Marlin controller and USG stay in sync?

BTW: Marlin 2.0 does support realtime position query with the right Configuration options.

#define M114_DETAIL         // Use 'M114` for details to check planner calculations
#define M114_REALTIME       // Real current position based on forward kinematics
 and use
M114RD

**Comment by mron** [comment]
I have a version of Marlin which supports GRBL single letter commands "!" and "?". I have a working quickstop that doesn't lose steps saves command and move buffers and restores them.
UGS is telling me "Grbl has not finished booting". What is it waiting for?

**Comment by winder** [comment]
Try adding the startup string:
https://github.com/winder/Universal-G-Code-Sender/blob/50e4bc471c1de429324a2e4e076411ea14bf74b6/ugs-core/src/com/willwinder/universalgcodesender/GrblUtils.java#L82

**Comment by andy-hnq** [comment]
> 
> 
> I have a version of Marlin which supports GRBL single letter commands "!" and "?". I have a working quickstop that doesn't lose steps saves command and move buffers and restores them.
> UGS is telling me "Grbl has not finished booting". What is it waiting for?

Which UGS branch are you running? The main branch or the Marlin branch? If you have the Marlin branch then you should be able to set the firmware dropdown to 'Marlin' (see screenshot). The branch has a class 'MarlinUtils' in the same package as 'GrblUtils' that recognises the Marlin responses.

![ugc-marlin-ss](https://user-images.githubusercontent.com/58275896/111032045-6be67b00-8402-11eb-8b8f-2edf61f18541.png)

It would also be good to know which Marlin version is the starting point for your changes.

I havent touched the UGS or Marlin code for a few months but if you are still having problems then I will dig into the code and try to figure things out. My most recent commit was a merge of changes from the main branch soon after your messages last September. This branch should be the best one to use...

https://github.com/andy-hnq/Universal-G-Code-Sender/tree/marlin-sep2020 

BTW, I am still using UGS+Marlin on my milling machine and it's still doing the job for me.

**Comment by mron** [comment]
Here's my Marlin fork:
https://github.com/mron/Marlin/tree/bugfix2.0.xautoreport. It has !, ? working. Restart also works with "R000". I'll add '~'. I'll try your version of UGS.

This is setup for a PC. How do I get it to run on a Mac?

**Comment by andy-hnq** [comment]
Java is cross platform so if you get the JDK for the Mac plus Maven (the build tool) then it should build and run. I have never used a Mac myself so can't do more than tell you the theory though.

**Comment by mron** [comment]
I'll try it out. The Marlin fork is raw. I've only tested it with simple, short samples. I'm trying to get UGS working so I can push it. If you try it let me know how it goes.

**Comment by andy-hnq** [comment]
Cool. It's probably best to just get the fork to build and talk to Marlin first before thinking about any changes.

I had a quick look at the code to refresh my memory...

The GRBL real time commands are sent by the following methods...

- com.willwinder.universalgcodesender.GrblController.requestStatusReport()
- com.willwinder.universalgcodesender.GrblController.pauseStreamingEvent()
- com.willwinder.universalgcodesender.GrblController.resumeStreamingEvent()

Then the replies are handled by com.willwinder.universalgcodesender.MarlinController.rawResponseHandler(String)

All of the above have direct equivalents in the MarlinController class

Currently the Marlin fork uses M114 for status checks (and only when Marlin is idle). This is done on a time interval basis from com.willwinder.universalgcodesender.MarlinController.createPositionPollTimer()

This is called by 
- com.willwinder.universalgcodesender.MarlinController.MarlinController(MarlinCommunicator)
- com.willwinder.universalgcodesender.MarlinController.rawResponseHandler(String)

The raw response handler also deals with the replies to the M114 commands.

HTH

I'll see if I can persuade your Marlin to install on my Arduino... but will need some time to refresh my memory about how to do that... it's been a while.

**Comment by mron** [comment]
Compare your Configuration.h and Configuration_adv.h with mine. I uploaded the config files for my machine, be aware of that. I think the main changes are in Configuration_adv.h. Look for:

#define EMERGENCY_PARSER

// Support realtime commands: S000 State, P000 Pause, R000 Resume
#define REALTIME_REPORTING_COMMANDS

// Send machine status reports while moving and status reports GRBL style
#define FULL_REPORT_TO_HOST_FEATURE

**Comment by mron** [comment]
Power just went out.
Look at Configuation_adv.h.
Look for:
#define EMERGENCY_PARSER
#define REALTIME_REPORTING_COMMANDS
#define FULL_REPORT_TO _HOST_FEATURE

**Comment by mron** [comment]
What command does UGS send while it’s waiting for the “Grbl” string?

**Comment by andy-hnq** [comment]
> 
> What command does UGS send while it’s waiting for the “Grbl” string?

It doesn't (with Marlin selected as the firmware). I ran it in the debugger just now to make sure. Marlin sends a bunch of info to GRBL when it boots. The first command sent by GRBL is M114 when the timer fires after a few seconds.

**Comment by mron** [comment]
Ok,
I’ll put that in.

**Comment by mron** [comment]
I got your UGS/Marlin running. I get a null pointer error when I connect.

java.lang.NullPointerException
	at com.willwinder.ugs.nbp.core.actions.ReturnToZeroAction.isEnabled(ReturnToZeroAction.java:76)
	at com.willwinder.ugs.nbp.core.actions.ReturnToZeroAction.lambda$UGSEvent$0(ReturnToZeroAction.java:69)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:764)
	at java.desktop/java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:717)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:711)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:89)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:734)
	at org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:136)
[catch] at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:199)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

I'll look into this. The USG console is showing what Marlin is sending out:

Grbl 0.9
Grbl 0.9
...

I picked version 0.9 out of the air. Would you show me what your Marlin is outputting at boot?

**Comment by mron** [comment]
I got your version of UGS/Marlin running. I get a null pointer error when I connect:
java.lang.NullPointerException
	at com.willwinder.ugs.nbp.core.actions.ReturnToZeroAction.isEnabled(ReturnToZeroAction.java:76)
	at com.willwinder.ugs.nbp.core.actions.ReturnToZeroAction.lambda$UGSEvent$0(ReturnToZeroAction.java:69)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:764)
	at java.desktop/java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:717)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:711)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:89)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:734)
	at org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:136)
[catch] at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:199)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

I'll look into this. I am seeing output on the console now:

Grbl 0.9
...

I picked 0.9 out of the air. Can you show me the boot output from your Marlin?

**Comment by mron** [comment]
My Mac has java version 9. Does that matter?

**Comment by mron** [comment]
Here's more info:
INFO [com.willwinder.universalgcodesender.model.GUIBackend]: Exception in openCommConnection.

**Comment by breiler** [comment]
I have not read through all the comments here but I noticed this in your branch @andy-hnq that would cause @mron's error:
https://github.com/andy-hnq/Universal-G-Code-Sender/blob/marlin-sep2020/ugs-core/src/com/willwinder/universalgcodesender/MarlinController.java#L71

You could just return the instance created here:
https://github.com/andy-hnq/Universal-G-Code-Sender/blob/97a069f94683f58cda0da83676c53457c2968800/ugs-core/src/com/willwinder/universalgcodesender/MarlinController.java#L32

The capabilities-object would need to be populated with the supported features to enable those buttons in the GUI:
https://github.com/andy-hnq/Universal-G-Code-Sender/blob/marlin-sep2020/ugs-core/src/com/willwinder/universalgcodesender/CapabilitiesConstants.java

**Comment by mron** [comment]
Thanks  for responding. I really want UGS and Marlin to work together. I'm pretty sure I can get the hooks that are in Marlin and add a few more to make this work. When working with Java, does the JDK version matter? I'm using JDK 9.

**Comment by mron** [comment]
BTW, Your suggestion worked!

**Comment by andy-hnq** [comment]
I'll have a look at the error later. I'm not sure what the minimum Java version will be for UGS but I suspect that JDK 9 will be OK.

The UGS Maven build seems to take a long time for me too. Part of the time is spent running unit tests. There are options to change what the script does but I haven't messed with them myself. I just go and make a drink while it does its thing. When messing with the code I use an IDE (Eclipse in my case but I think UGS is more set up for NetBeans). The IDE just builds the code that I changed and I can run and debug the thing straight away.

**Comment by andy-hnq** [comment]
> I got your UGS/Marlin running. I get a null pointer error when I connect.
> 
@breiler is correct - the getCapabilities method is returning null. It was auto generated by the IDE and I have not written a more sensible version. I will do so.

But... I'm not seeing an NPE when I connect. 

I will look further but it does occur to me that serial comm support is one of those areas where the cross platform guarantee may run out so this could be related to running on a Mac - I'm not saying that is certain but it is a possibility. It's not something I can check out but maybe someone else has experience of running UGS on a Mac?

Normally UGS would complain that it cant find any comm ports though... just thinking out loud.

> I picked version 0.9 out of the air. Would you show me what your Marlin is outputting at boot?

I copy/pasted the text below from my UGS window after connecting. I have not messed with Marlin to make it 'pretend' to be GRBL - it's just the standard Marlin output. The M114 is sent a few seconds after boot. The M114 is not critical to making it work - it just requests Marlin to report the position so that UGS can update the UI fields.

```
**** Connected to COM3 @ 115200 baud ****
start
echo:Marlin bugfix-2.0.x
echo: Last Updated: 2019-11-18 | Author: (none, default config)
echo:Compiled: Dec  1 2019
echo: Free Memory: 5786  PlannerBufferBytes: 1120
echo:Hardcoded Default Settings Loaded
echo:  G21    ; Units in mm (mm)
echo:Filament settings: Disabled
echo:  M200 D3.00
echo:  M200 D0
echo:Steps per unit:
echo: M92 X200.00 Y200.00 Z200.00 E500.00
echo:Maximum feedrates (units/s):
echo:  M203 X100.00 Y100.00 Z40.00 E25.00
echo:Maximum Acceleration (units/s2):
echo:  M201 X200.00 Y200.00 Z100.00 E10000.00
echo:Acceleration (units/s2): P<print_accel> R<retract_accel> T<travel_accel>
echo:  M204 P200.00 R3000.00 T400.00
echo:Advanced: B<min_segment_time_us> S<min_feedrate> T<min_travel_feedrate> J<junc_dev>
echo:  M205 B20000.00 S0.00 T0.00 J0.01
echo:Home offset:
echo:  M206 X0.00 Y0.00 Z0.00
echo:PID settings:
echo:  M301 P22.20 I1.08 D114.00
echo:Backlash compensation:
echo:  M425 F0.00 X0.00 Y0.00 Z0.00
>>> M114
[verbose] X:0.00 Y:0.00 Z:0.00 E:0.00 Count X:0 Y:0 Z:0
ok

```

**Comment by mron** [comment]
That's interesting. I just pushed a version that pretends to be GRBL. I put that in before I used @breiler fix. I'll take it out and see if I connect. The '!' has a slightly different output than M114.
[verbose] X:0.00 Y:0.00 Z:0.00 E:0.00S_XYZ:3
S_XYZ:# is supposed to be the grbl status number. It's very unreliable right now. It's something I'll fix.

**Comment by mron** [comment]
I took out the fake grbl message. USG connected fine.

**Comment by andy-hnq** [comment]
Cool. Can you run G code and move steppers now?

**Comment by mron** [comment]
Yes!

**Comment by andy-hnq** [comment]
Excellent. I dont yet have my Arduino re-flashed but will look at that later on. Need to take care of some household tasks for the next few hours.

**Comment by mron** [comment]
I meant '?' should work any time

**Comment by andy-hnq** [comment]
Yes. I get that... I need to get your Marlin onto my Arduino before I can offer much more help but I wont have chance to look at that until later.

**Comment by mron** [comment]
If you have time, I'd like to work together on this. I think we can progress pretty fast. I'm pretty free right now, let me know when you have time.

**Comment by andy-hnq** [comment]
Yes, no problem at all, subject to real life priorities, day job, sleep etc :)

**Comment by mron** [comment]
Go to it! Thanks

**Comment by mron** [comment]
I tried building with jdk-11 on my mac. I'm using VSC for my IDE, jdk-11 seems to be it's preference. I don't have a problem using jdk-9, I wanted to try it. I did a clean build with 11.  UGS crashes:

[INFO] Compiler: HotSpot 64-Bit Tiered Compilers
[INFO] Heap memory usage: initial 64.0MB maximum 4096.0MB
[INFO] Non heap memory usage: initial 7.3MB maximum -1b
[INFO] Garbage collector: G1 Young Generation (Collections=31 Total time spent=0s)
[INFO] Garbage collector: G1 Old Generation (Collections=0 Total time spent=0s)
[INFO] Classes: loaded=7912 total loaded=7912 unloaded 0
[INFO] INFO [org.netbeans.core.ui.warmup.DiagnosticTask]: Total memory 17,179,869,184
[INFO] INFO [null]: Total physical memory 17,179,869,184
[INFO] 2021-03-14 07:04:15.906 java[61949:2569825] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'NSWindow drag regions should only be invalidated on the Main Thread!'
[INFO] *** First throw call stack:
[INFO] (
[INFO] 	0   CoreFoundation                      0x00007fff36edfa9d __exceptionPreprocess + 256
[INFO] 	1   libobjc.A.dylib                     0x00007fff616d53c6 objc_exception_throw + 48
[INFO] 	2   CoreFoundation                      0x00007fff36ef98f5 -[NSException raise] + 9
[INFO] 	3   AppKit                              0x00007fff344905ca -[NSWindow(NSWindow_Theme) _postWindowNeedsToResetDragMarginsUnlessPostingDisabled] + 317
[INFO] 	4   AppKit                              0x00007fff3448d9f7 -[NSWindow _initContent:styleMask:backing:defer:contentView:] + 1479
[INFO] 	5   AppKit                              0x00007fff3448d42a -[NSWindow initWithContentRect:styleMask:backing:defer:] + 45
[INFO] 	6   libnativewindow_macosx.jnilib       0x000000012aa293fe Java_jogamp_nativewindow_macosx_OSXUtil_CreateNSWindow0 + 398
[INFO] 	7   ???                                 0x000000010d715690 0x0 + 4520498832
[INFO] )
[INFO] libc++abi.dylib: terminating with uncaught exception of type NSException
[INFO] /Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/ugsplatform/bin/../platform/lib/nbexec: line 425: 61949 Abort trap: 6           "/Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home/bin/java" -Djdk.home="/Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home" -classpath "/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/ugsplatform/platform/lib/boot.jar:/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/ugsplatform/platform/lib/org-openide-modules.jar:/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/ugsplatform/platform/lib/org-openide-util-lookup.jar:/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/ugsplatform/platform/lib/org-openide-util-ui.jar:/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/ugsplatform/platform/lib/org-openide-util.jar" -Dnetbeans.dirs="/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/ugsplatform/ide:/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/ugsplatform/platform:/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/ugsplatform/java:/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/ugsplatform/ugsplatform:/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/ugsplatform/extra:" -Dnetbeans.home="/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/ugsplatform/platform" '-Xdock:name=ugsplatform' '-Xdock:icon=/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/ugsplatform/bin/../../ugsplatform.icns' '-Xms64m' '-Xverify:none' '-Dsun.java2d.noddraw=true' '-Dsun.awt.noerasebackground=true' '-Dnetbeans.indexing.noFileRefresh=true' '-Dnetbeans.logger.console=true' '-ea' -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath="/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/userdir/var/log/heapdump.hprof" org.netbeans.Main --cachedir "/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/userdir/var/cache" --userdir "/Users/ron/Downloads/Universal-G-Code-Sender-marlin-sep2020/ugs-platform/application/target/userdir" "--branding" "ugsplatform" "--branding" "ugsplatform" 0<&0
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  25.814 s
[INFO] Finished at: 2021-03-14T07:04:15-07:00
[INFO] ------------------------------------------------------------------------

**Comment by andy-hnq** [comment]
That looks like a native code exception. It should not happen even if the Java code does the wrong thing, so possibly a bug in that version of Java on the Mac. TBH if Java 9 is working then I'd stick with that - pick your battles.

**Comment by mron** [comment]
Agreed, I have enough to do. I'm trying to add feedrate to the DRO status. I'll look at the visualizer update next, see if can make sense of it.

**Comment by mron** [comment]
The way Marlin state is handled doesn't match well to Grbl. I'm going to separate them in Marlin.

**Comment by mron** [comment]
I see why the status timer doesn't start. It's looking for a string "Free Memory:".  That doesn't show up.

**Comment by andy-hnq** [comment]
Just pushed a commit containing some code to handle axis reset... not related to the current realtime command discussion but I should have committed this stuff months ago

I have now flashed my controller with a version built from your git fork. It took me a while to remember how to redo my pin config but that's all working now and is moving steppers when requested by UGS. Don't get too happy yet...

Then I realised that your realtime stuff was not present. Git clone had put me on the 2.0.x branch. I checked out the bugfix2.0.xautoreport branch and tried to build again. I have a compilation error in the Arduino IDE...

```
sketch\src\lcd\extui\lib\dgus\mks\DGUSScreenHandler.cpp:23:10: fatal error: ../../../../inc/MarlinConfigPre.h: No such file or directory

 #include "../../../../inc/MarlinConfigPre.h"

          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

compilation terminated.

exit status 1
Error compiling for board Arduino/Genuino Mega or Mega 2560.
```
Maybe I'm missing something simple but this doesn't make sense to me just now - the path in the include statement looks OK but the directory path element 'sketch' is not part of the source tree so maybe the Arduino IDE is doing something more complicated?

**Comment by mron** [comment]
I'll try changing the board and compiling here.

**Comment by mron** [comment]
I'm seeing issues, but I think it's related to the toolchain. The atmel compiler vs arm. I'll work on this. Are you using VSC and Platformio?

**Comment by mron** [comment]
Oops, I guess I should be using a different IDE. I've never built for Atmel. This will take me while.

**Comment by andy-hnq** [comment]
I just open the .ino file in the Arduino IDE and hit the build button. Not sure how that works under the hood. I can see that it's running g++ processes but beyond that I have not figured out how it works. Maybe the 'sketch\' path is a clue that it is using a temporary directory but I have not found it so far.

**Comment by andy-hnq** [comment]
I found the 'sketch' directory now - it's in a folder 'arduino_build_426964' under my temp directory. The folder is cleared when I exit the IDE so my theory that old build files from other branches might need cleaning up does not hold water. 

I will try installing the latest IDE version but that won't be today.

**Comment by mron** [comment]
Marlin wants 1.8.8 or better. I’m downloading 1.8.10 now. I’ll let you know if it works.

**Comment by mron** [comment]
There is a "../" missing in the file indicated. There are a couple c++ errors I'm running down now.

**Comment by andy-hnq** [comment]
I was on 1.8.10. I have now installed 1.8.13. Same issue I'm afraid.

**Comment by mron** [comment]
Yes, Sorry for the trouble. I'm running into 2 different problems depending on the IDE I use. Platformio has a linking problem, Arduino IDE has different issues. This is going to take a while.
I'll let you know when I sort it out.

**Comment by andy-hnq** [comment]
No worries. 

FYI I tried building the latest bugfix branch from the main Marlin repository. That does not hit the compile error but it won't link with the Arduino IDE... 'avr-gcc.exe: The filename or extension is too long'. I'm not sure if it is carping about a single file name or just that the list of .o files being linked is too long. I do see other folks reporting this error with various versions of Marlin but have not found a clear (and relevant) answer about a fix. Lots of folks suggest moving the source tree nearer to the top of the file system but that's not where the IDE does the build so that won't help.

I guess that if we can get past the compile issue with your branch then we may still hit the linker issue.

My previous build was also from the bugfix branch. It built just fine but that was in late 2019 and a lot has changed since then.

**Comment by mron** [comment]
I have compiled it with Platformio. I'll try it on an Arduino before I release it to you. I'm still working on the Arduino IDE stuff. Interesting since they use the compiler.

**Comment by mron** [comment]
The Marlin group has been very busy. I'm trying to get the CNC support noticed so it can get into the main bugfix branch. One of the group wrote a very cool and quite accurate 3D printer simulator. I used to develop the feedhold stop. It made tracking the action of the command queue , planner and stepper possible.

**Comment by mron** [comment]
I found the problem with Arduio IDE. I'd left an unfinished stub. I forgot about it. Platormio didn't care for reason. I got rid of it and it compiles. I want to run minimal tests on it before I turn it over.

**Comment by mron** [comment]
I loaned my Arduino board to a friend, I don't think I have another. I'll pick it up later today and test it.

**Comment by andy-hnq** [comment]
I have now refreshed my UGS fork from the main repository. I've run a couple of XY axis only test jobs from the classic UI with a plastic pointer in the mill. It seems sane. The controller is running the build of Marlin 2.0.x that I compiled yesterday.

The new branch is...

https://github.com/andy-hnq/Universal-G-Code-Sender/tree/marlin-mar2021

I needed to disable two test cases in S3FileSystemViewTest to get the mvn build to complete - this one has been failing for some time for me but I've been nowhere near that code. I see there is a recent commit in the master branch relating to failures of this test but mine still failed after merging.

**Comment by mron** [comment]
I overrode the open/close comm ports on my fork of UGS. I start the polling timer as soon as a connection is made and stop it when port is closed. This works better for the Mac.

**Comment by mron** [comment]
I made a pull request for your mar2021. I'm not clear about pull requests. Am I asking you to incorporate my changes, or am I asking you permission to incorporate your changes. I want to get your changes into my fork

**Comment by mron** [comment]
I have a question about the status request timing. The status request timer is independent of the code sender. Can the status request come in the middle of a command? Can the controller get something like "G1X0.?1Y0Z5.5F100" ?
Also, I think I understand why the status response needs to be delimited in a unique way, like <<X:0.00 Y:1.5 ... >> and it should not send an "ok". I think the ok's are confusing UGS knocking it out of sync.

**Comment by andy-hnq** [comment]
I used this article... https://gist.github.com/CristinaSolana/1885435 . You want an upstream fetch I think, not a pull request. Then you may get some merge conflicts... it depends which files have changed in your branch. On github you can also compare branches across related forks... for example I compared my mar2021 branch with the main repository master branch after I'd merged the latest changes. I won't pretend that I found it an easy process and there were some unexpected turns but I think the end result was what I intended.

I don't know the answer to your last question sorry - I guess you need to know whether it can happen with UGS+GRBL and also what is possible with Marlin. I'm not sure there is much benefit in being able to do it though - most g code commands are very short so sending them should be fast. Repeatably testing this kind of thing is also tricky. Executing commands on the other hand is not quick so being able to update the position display in mid move or while working through a backlog of multiple commands is a big benefit. 

What's the URL for your fork BTW?

Are you using the classic UI or the platform UI to test with?

**Comment by mron** [comment]
Here's my fork: https://github.com/mron/Universal-G-Code-Sender/tree/marlin-mar2021
I'm not good with git. I get confused about the terms. I made a new fork of yours and updated marlin-mar2021 with my stuff. I haven't run it yet, but you can see what I'm trying to do.

**Comment by mron** [comment]
I'm trying to configure VSC to work with the java fork. Another rabbit hole, I want to compile in the VSC IDE. This could take a while, I can't even get "Hello World" working.

**Comment by mron** [comment]
I am making progress. I found a bunch of files containing "public static void main". "MainWindow.java", "VisualizerCLI.java" ... I can run|debug most of them. They are UI elements. Where is the main that pulls them together into a single UI? I'm still looking.

**Comment by andy-hnq** [comment]
I have used the platform UI very little... practically not at all so far. I have a quick code/debug cycle working with the classic UI and the classic UI is the one I have used to cut metal. For the classic UI you need the ugs-classic project and the main class is...

com.willwinder.universalgcodesender.MainWindow

It's all pretty straightforward. 

I thought since you have been using the platform UI then I really should put some effort into it myself. I've had a bit of a nightmare getting a working build (not sure why - best guess is it was conflicting in the Maven repos with a downloaded UGS binary version from the main web site). 

I've also been struggling with the debug setup for the platform UI. There are some suggestions in the readme at the root of the git tree which may help... but I dont have that working properly at the moment. I finally got a remote debug setup to work. After a lot of failed attempts I finally found that adding this line to the configuration section of the nbm-maven-plugin in ugs-platform/pom.xml worked...

<additionalArguments>-J-Xrunjdwp:transport=dt_socket,server=y,suspend=y,address=5005</additionalArguments>

This means you run the same way as normal. It pauses so you can connect the debugger (IDE) to port 5005.

HTH!

I have now just run a test g code job on the machine from the platform UI. Seems somewhat sane but I could not find any way to resume after pausing in the G code, so there is stuff to do still.

**Comment by mron** [comment]
Thanks Andy,
I think the MainWindow.java has everything I need to debug the Marlin comms. Lots faster than mvn from the command line! I hope to have the Marlin side ready for you soon.

**Comment by mron** [comment]
Ok, I have a version that compiles in Arduino IDE and uploads:
https://github.com/mron/Marlin/tree/bugfix2.0.xautoreport
The Configuration.h is setup for an arduino.

Here's the UGS marlin. It's forked from your marlin-mar2021, but I made a mess of the commits. I'll try and fix it.

https://github.com/mron/Universal-G-Code-Sender/tree/marlin-mar2021

**Comment by mron** [comment]
The two work. There are definitely issues.

**Comment by andy-hnq** [comment]
I've been testing some more today by writing a unit test and trying to make MarlinController less stubby.

I've found a couple of anomalies with pause/resume handling which I am trying to resolve. I was hoping this would fix the random behaviour of the pause/resume button on the classic UI when it (and Marlin) process M0 in the g code. Right now my unit test cases are passing but the more 'rational' behaviour seems to work less well in the UI, not better.

This is all related to the active command lists and the ability for some commands to queue jump so I really need to get M0/M118 stable before doing anything else.

**Comment by mron** [comment]
I forgot to put feedhold into M0! I only put it in ‘!’. I’ll push a fix ASAP. What is M118?

**Comment by andy-hnq** [comment]
M118 was a typo. I meant M108... which is what I send to resume after M0

**Comment by mron** [comment]
Ok, I just pushed them. I’ve not tested yet

**Comment by mron** [comment]
Andy,
I just realized M0 is not caught by the emergency parser. It’s treated like any other command. I need to fix that. I have to go out a bit, should be back in a couple hours.

**Comment by mron** [comment]
I pushed a version of Marlin with the M0 implemented in the e_parser. Seems to work on ugs-platform. If you haven't tried Marlin support in Platformio, take a look at it. It's much faster than Arduino IDE for compile and loading. Much more modern toolset.

**Comment by andy-hnq** [comment]
I have just pushed a commit (to the same branch). I have added some unit tests for the controller, made the controller less stubby and spent some time looking at the pause button behaviour and program row count with the classic UI. Both now work better and my XY axis test job works on the mill.

The test job also works on the platform UI BUT only if I remove the M0 command, otherwise M0 is sent, then Marlin replies 'echo:busy: paused for user'. This puts UGS into 'HOLD' state and the pause button (which was enabled) is disabled. 

I have not used the platform UI much. Can anyone advise what should happen when the G code program contains an M0. Should the pause button be enabled?

**Comment by mron** [comment]
I’m working on the controller side. The status isn’t updating when paused. I’m running that down now.

**Comment by mron** [comment]
I’m staying off ugs-platform until I get everything in MainWIndow.java working smoothly.

**Comment by andy-hnq** [comment]
There are changes in my commit around pause/resume. In some situations isPaused() was returning the wrong answer. Another issue was that the controller object was correctly in a paused state sometimes but was not sending notifications to the UI that the state had changed - there are multiple types of notification.

BTW, I checked the behaviour when running the platform UI - isPaused() is being called plenty of times and is reporting the correct answer so there is more work needed to find out why it works in the classic UI and not in the platform UI.

Note also that 'status' and 'state' are different things in the UGS controller classes. Also there are multiple places where g code commands are stored. They all need to fly in formation...

- com.willwinder.universalgcodesender.BufferedCommunicator.activeCommandList (queue)
- com.willwinder.universalgcodesender.BufferedCommunicator.nextCommand (single ref)
- com.willwinder.universalgcodesender.BufferedCommunicator.commandStream (stream reader)
- com.willwinder.universalgcodesender.BufferedCommunicator.commandBuffer (queue)

- com.willwinder.universalgcodesender.AbstractController.activeCommands (list, private)
- com.willwinder.universalgcodesender.AbstractController.streamCommands (stream reader)

**Comment by mron** [comment]
I just pushed another Marlin. The status switches to HOLD when I pause. I’m going to switch to your latest mar2021.

**Comment by andy-hnq** [comment]
> The status switches to HOLD when I pause. 

Same here but the label and enabling of the pause/resume button is driven by a different API call and pub/sub notification so it can show HOLD and the pause/resume button can be disabled or have the wrong label. I think my commit yesterday has sorted the behaviour in the classic UI but the disabling of the button in the platform UI is still a problem and I don't yet know why.

I'm looking again at installing your Marlin fork. The compile issue is resolved but as I half expected, the Arduino IDE build fails during linking. I have one more thing to try and if that doesn't work I'll try the other tool chain.

**Comment by mron** [comment]
I’m up late.I’m trying a few things. Hold off installing till tomorrow.

**Comment by mron** [comment]
Hi Andy,
I spent hours trying to figure out why commands like M0, M108 aren’t sending ‘ok’ when the commands are captured by the e_parser. I really don’t like the way it’s been implemented in Marlin.

Please switch to single byte commands for status, pause and resume. Use ’sendByteImmediately’. This sends the byte to the controller with no expectation of response. I pushed a version of your UGS that implements it. 
I also pushed a new version of Marlin. Here’s a example of a byte send.

    protected void sendStatus() throws Exception {
        if ( this.capabilities.hasCapability(GrblCapabilitiesConstants.REAL_TIME)) {
            logger.info("trying to '?'");
            this.comm.sendByteImmediately(MarlinUtils.GRBL_STATUS_COMMAND );
            //nonScriptedCommandsSent++;
        } else {
            logger.info("trying to send ? as command");
            GcodeCommand command = createCommand("?");
            // sendCommandImmediately(command);
            this.marlinComm.sendRealtimeCommand(command);

            // adjust the sent counter
            nonScriptedCommandsSent++;
        }
    }

**Comment by andy-hnq** [comment]
I'm stuck just now with building your fork. I've installed VSC and platformio but I use a custom pin mapping and it doesn't like it. It complains that my definition of MOTHERBOARD is not compatible with my selection of mega2560 as the env in platformio.ini. I know my pins and boards.h stuff is OK but platformio is being clever and using additional metadata which does not know about my custom pin setup.

The docs suggest adding a new board type using a json file in buildroot/share/PlatformIO/boards. It says 'use an existing one as an example' but I dont see anything in there that relates the MOTHERBOARD value from boards.h to the board name used by platformio, nor can I see an example there that looks like a good template for an Arduino Mega.

With M0/M108 I *DO* get 'OK' but it's a bit non intuitive. There is no 'ok' straight after 'M0'. Instead the controller repeatedly sends 'echo:busy: paused for user'. When M108 is sent to resume processing, *TWO* 'OK' messages appear.

I'd like to do a version of UGS that will work either with or without the single byte commands. Ideally it would be good to be able to detect whether they are available by looking at the startup messages.

**Comment by mron** [comment]
I'll keep working on the 'ok' problem. The pins stuff is spread out. What pins are you needing?

**Comment by mron** [comment]
Try starting with a generic RAMPS CNC, I set MOTHERBOARD to BOARD_RAMPS_14_SF
In pins_RAMPS.h
//
// M3/M4/M5 - Spindle/Laser Control
//
#if HAS_CUTTER && !defined(SPINDLE_LASER_ENA_PIN)
  #if !NUM_SERVOS                                 // Use servo connector if possible
    #define SPINDLE_LASER_ENA_PIN              4  // Pullup or pulldown!
    #define SPINDLE_LASER_PWM_PIN              6  // Hardware PWM
    #define SPINDLE_DIR_PIN                    5
  #elif HAS_FREE_AUX2_PINS
    #define SPINDLE_LASER_ENA_PIN             40  // Pullup or pulldown!
    #define SPINDLE_LASER_PWM_PIN             44  // Hardware PWM
    #define SPINDLE_DIR_PIN                   65
  #else
    #error "No auto-assignable Spindle/Laser pins available."
  #endif
#endif

I was able to buildin Arduino as well as Platformio. Please send Arduino error output.

**Comment by andy-hnq** [comment]
I use a Protoneer CNC shield. It is intended for GRBL on an Uno and uses the GRBL pins. GRBL on the Mega can't use those pins but Marlin can.

It was the 'file name too long' error. The file names are fine but there are too may of them and they blow a limit (32K chars maybe). It may be a Windows specific issue but plenty of people have hit the same thing.  I tried moving my TMP dir but that didn't help because the file names in the link command are all relative paths so moving the TMP location makes no difference to the length of the link command string.

I'm not against using VSC and platformio... if I can figure out how to make it understand that my custom motherboard #define is compatible with the Mega.

**Comment by andy-hnq** [comment]
Found it. 

I decided that it must be something specific to Marlin and I was right... Marlin/buildroot/share/PlatformIO/scripts/preflight-checks.py . This does a text scan of Marlin/src/pins/pins.h . 

It requires a specific line comment format after each #include and derives the metadata from this. Once I added the line comment after my extra #include all was well. No need for a custom board type.

I now have your version on the controller and have given it a quick try.

I found two issues running my first test script...

```
G21 (Units in millimeters)
G90 (Absolute programming)
G0 X5 F40
M0 (PAUSE)
G0 X0 F40
```

Firstly the pause happens before the first move has completed. This isn't what other firmwares do (GRBL included IIRC). I often use M0 with a two axis job to prompt myself to manually turn the spindle on or off, raise or lower the Z axis or drill a hole.

Secondly after M0/M108 there is only one 'OK'. This is one fewer than the master version of Marlin so UGS never sends the remainder of the script and remains stuck in RUN mode.

**Comment by mron** [comment]
Great! Do you think I should have M0 wait until move is complete? The reason I took this on was to be able to feedhold in the middle of a move. It’s much easier to have an M0 hold at the end of a move, and ! hold in the middle of a move. I don’t know how they “supposed” to behave.

**Comment by andy-hnq** [comment]
https://marlinfw.org/docs/gcode/M000-M001.html

"The M0 and M1 commands pause _after the last movement_ and wait for the user to continue"

So yes, it should wait until the previous move is complete.

The single byte commands should be real time.

**Comment by mron** [comment]
I pushed a version that reverted M0/M108 to normal. Try it. I have to go a out for a few hours. I'll look it then.

**Comment by mron** [comment]
I have a feeling M108 might need to stay in the emergency parser.

**Comment by andy-hnq** [comment]
I pushed a commit to try out '?'.

I haven't found a nice way to turn realtime behaviour on and off so I have used a boolean constant in the MarlinController class until there is a better way.

A few testing observations...

- Reset zero doesnt work with REALTIME_ENABLED. The command is sent to Marlin but the status strings coming back show the coordinates dont go back to zero.
- The trivial script (same as before but without the M0) doesn't end with REALTIME_ENABLED. I havent really looked into why at the moment.
- I also noticed that with REALTIME_ENABLED set to false the M114s report fractional values during the job. Probably fine.
- I also noticed that '?' responds with 'OK' if I type it in the command entry field - probably a good thing.
- The status output seems to appear in the window regardless of 'verbose'. One more to figure out if we want more frequent position updates.

**Comment by mron** [comment]
G92 will take a bit of work. The status command, and M114 I think, return the stepper step count converted to mm. Marlin has the functions to convert to work coords, I'll look at them later. I rushed out that last Marlin update. I took feedhold out of the e_parser.h, but forgot to put it back into M0_M1.cpp. I fixed that and pushed it, again without testing. I'll test it and let you know.
I want to work with gcode flow. When I feel like it's working reliably, I'll look at G92.
I agree about the 'verbose', I think it might be set with Logging level.

**Comment by andy-hnq** [comment]
OK take your time. I have plenty to look at on the UGS side and not sure how much free time I will have.

**Comment by andy-hnq** [comment]
I've been looking a bit more at the real time way of working. UGS has two ways of working with GRBL. 

With non-realtime it works by counting the number of commands sent and 'ok's received back. Once it has an 'ok' for every command in the script it pops up the dialog to say that the script has finished and goes back to IDLE.

With realtime this behaviour is overridden in GrblController so that the replies to '?' control the state of UGS... so GRBL becomes the main factor in deciding the state reported by UGS.

I dug out an Uno with Grbl to get an example. The output below is running G0X20F40 with me sending a couple of '?'s via the Arduino serial monitor (which also sends a line break)

```
Grbl 1.1h ['$' for help]
ok
<Run|MPos:12.825,0.000,0.000|FS:500,0|WCO:0.000,0.000,0.000>
ok
<Idle|MPos:20.000,0.000,0.000|FS:0,0|Ov:100,100,100>
ok
```
Unlike Grbl, your response to '?' doesn't have any indication of state. 

Like Grbl, your fork replies 'ok' to each command before the move has completed. This means that counting 'ok's wont work for tracking script progress... which is probably why Grbl replies with the state info.

One more worry for me... I tried having UGS send multiple commands and having Marlin buffer them a while back (much higher up this thread) and it resulted in the job going wrong and Marlin incorrectly reporting the position at the end of the job. Maybe this has been resolved now but I never really pinned it down so AFAIK it was not a tracked issue.

TBH I had not realised that the single byte commands were quite so coupled to changes in the 'ok' reply semantics. Sorry - my bad.

**Comment by andy-hnq** [comment]
I think my previous comment about 'ok' semantics having changed was incorrect. The 'ok' semantics are not as simple as I had thought - Marlin can buffer commands but it seems not many before it reports 'busy: processing'.

I've now pushed a commit which separates out the realtime command support into a subclass. This can be selected as the firmware before opening the connection. 

I've tested the non-realtime support with my test job and the 2.0.x branch of Marlin (main repos - reports version 2.0.7.2) and it runs the job correctly, including an M0 pause. It also detects job completion.

I also tested the realtime option with your branch...
-  It runs the job (without the M0) on my mill and visually it seems to follow the correct path.
- The status updates seem to work but...
- I do notice an audible glitch in the stepper movement each time a status query happens. 
- Job completion is not detected.
- Marlin reported the position as X and Y 0.09 at the end - the last command in the job is an absolute move to 0,0 so this is pretty odd.

The single byte pause/resume commands are now being sent if the 'Marlin Realtime' firmware is selected in UGS. I have not tried them much, although resuming from M0 did not work.

All of the above was done with the classic UI.

I will look some more at job completion detection.

**Comment by mron** [comment]
Thanks for separating out the real time stuff. I’ll take a look at how you did it.
My Internet is down. I won’t be able to pull or push anything until Monday or Tuesday

**Comment by mron** [comment]
The audible glitches were noted by other people. It seems more a problem on Arduinos.

**Comment by mron** [comment]
I pushed a Marlin that works well with my realtime fork of UGS-Marlin. Workspace coordinate and Machine space are in the status field. I changed the UGS command buffer to 1. This solved a problem I don't understand. When there are commands in the buffer, UGS will send them, even when paused.
I think I need to sync the '!' feedhold to the beginning or end of commands.

**Comment by Trakyan** [comment]
Is there any precompiled version that supports marlin? I've been trying to compile for the last few hours, but I can't seem to get java or mavin running let alone compile UGS. This seems a bit beyond me.

**Comment by mron** [comment]
I had to cook up my own version of marlin to make this work. Vanilla Marlin lacks a few thing's to make it work well. 
You can save yourself a lot of grief by setting up a GRBL controller.

**Comment by Trakyan** [comment]
I dont really need a full feature set, just for ugs to connect, jog the axis and send premade gcode files. I dont think I can use grbl, it lacks a few features that i need such as two motors and two endstops for one axis, custom mcodes (i havent been able to find anything on this online with grbl) and a few other features. Im actually running reprap firmware, but it can mimic marlins output, which i figure is my best bet since support for reprap firmware outside of 3d printing applications seems very thin. I was using laserweb, but jogging doesnt seem to work with the new reprap firmware i "upgraded" to. I should have stuck to the smoothie controller and hacked together a way to home two y axis motors.

**Comment by mron** [comment]
I’m not familiar with smoothie. Marlin had a viscous learning curve. Lots of community support, but brutal for me. The biggest problem I had, and still have is the heavy use of macros to help support all the variants. Klipper?

**Comment by mron** [comment]
Have reached out to the GRBL community about dual motors?

**Comment by Trakyan** [comment]
I found smoothieware pretty simple, but I don't have very complex requirements. Just a few custom mcodes for turning things on and off. The basic config like steps/mm, acceleration, pinouts etc. Was all super straightforward. Havent touched marlin because I've heard its quite difficult and cumbersome to configure. Reprap firmware I would not reccomend personally, a lot of the way it runs seems backwards to me and isnt suitable for things other than 3d printers (only g0-g3 are buffered so m codes and other gcodes run out of order to how they appear in the file.).

Ive looked through other forum posts regarding dual motors on grbl. The usual answer is jumper the step /dir togethrr which doent let them move independently, questions asking about two motors with two endstops either say its not supported or go unresolved.

**Comment by hm2dev** [comment]
After having read through the whole thread, are these two the latest works in terms of adding support for Marlin?
https://github.com/fedetony/Marlin/compare/bugfix2.0.xautoreport...mron:Marlin:bugfix2.0.xautoreport
https://github.com/winder/Universal-G-Code-Sender/compare/master...andy-hnq:Universal-G-Code-Sender:marlin-sep2020

Any chance of getting the things that are/were already working into the main branch, maybe as "experimental firmware" option or is Marlin support completely abandoned?

It would be really great to get at least the Jog Controller and sending single G-Code commands working with Marlin.

**Comment by breiler** [comment]
I experimented with Marlin in [this branch](https://github.com/winder/Universal-G-Code-Sender/compare/master...breiler:Universal-G-Code-Sender:feature/marlin) a long time ago. In my opinion the firmware is missing some key features to properly integrated with with a gcode sender. What are you hoping to do with it?

**Comment by hm2dev** [comment]
Thanks for the fast reply.
I was mainly looking for a way to get keyboard or even joystick based jog control for an universal positioning system which is driven by a Rambo und Marlin based driver board.
I don't want to change the setup itself as it is part of a larger toolchain, so abandoning Marlin is currently not really an option.
While looking for a way to avoid writing/hacking something myself I stumbled upon UGS, which looked like it would be the perfect solution, being able to send gcode files generated from my toolchain and providing manual jog control.
Thats until I found, that there is no support for Marlin.
What are the critical missing parts from Marlin?
From what I got from the above, (one of) the problems used to be a way to asynchronously obtain the current position, which seems to be implemented now with the M114_REALTIME option, or even M154 for auto position updates, enabled by setting EXTENDED_CAPABILITIES_REPORT.

**Comment by breiler** [comment]
I don't remember all problems I ran into, but reporting was one of them. Even if Marlin comes with an auto report feature now, it needs to be compiled in. And there is no way for UGS to know if it is - except for maybe query for the capabilities using the [M115](https://marlinfw.org/docs/gcode/M115.html) command. But in order for that to work that feature  needs to be compiled in as well. 🤦🏻‍♂️

That is just a small example of the problems we would run into. I felt at the time that Marlin is just not easy to integrate with and it would not be reliable enough.

You can have a look at https://cnc.js.org/ which should have some Marlin support.

**Comment by agowa** [comment]
@breiler what happens if you just send that command when it is NOT compiled in? Does the controller firmware hang, return no data, or return wrong data?

If it hangs the firmware doing some kind of 1st time "identifying your controller"-setup followed by an "please power off and on your controller now" in case it crashed the firmware.
No data would be easy, the only thing that could cause problems is when it sends wrong data that looks right.

Also would if Marlin firmware support is not being added in the near future, would it at least be possible to identify the controller having Marlin firmware and showing an error message? I scratched my head for weeks why universal gcode sender didn't work but octoprint did...

**Comment by dnikodem84** [comment]
I was recently donated Mini-Rambo 1.3 board ( https://ultimachine.com/products/mini-rambo-1-3 ), configured for a "V1 Engineering Lo-rider CNC V2" ( https://docs.v1e.com/lowrider/lrv2/indexv2/ ) 

I am rejuvenating , [This Changeset](https://github.com/winder/Universal-G-Code-Sender/compare/master...breiler:Universal-G-Code-Sender:feature/marlin) as a starting point and hoping to get enough machine control via the UGS to allow me to finish assembly and start using the machine. 

@breiler , I am wondering what your primary concerns with me creating a PR once I have confirmed I can speak to my new controller and have passed the Unit tests ?

**Comment by breiler** [comment]
Go for it. It was a long time ago I looked at this and a lot has happened in the UGS backend  since then (and in Marlin) which could make things simpler to implement. Might be easier to start with a fresh branch.

My main concern is still the possibility to be able to query the controller for its capabilities. What is the minimal requirements in Marlin to be able to communicate with it? 

You have the checklist in my original post with basic features that is expected to work (in some form or another).

**Comment by Plasma2025** [comment]
Hi, I use CNC3018 machine under Marlin 2.0 firmware. For a long time I'm trying to found any solution for PCB surface sensing before milling (autoleveling). Unfortunatelly without any result. There are many software like Candle, which support Autolevel for GRBL, but not for Marlin. I found that CNCjs supports Marlin, but Autolevel widget doesn't provide autoleveling future for Marlin. So, please inform does Universal G-codeSender supports autoleveling for Marlin?
